### ISSUE TITLE
SimpleVideoFileFilter example in Swift

### ISSUE THREAD

**Issue by ignasivegas** [BradLarson/GPUImage#2240]
Hi,

I'm trying to reproduce the SimpleVideoFileFilter example in Swift, everything goes perfect until I want to implement the setcompletionBlock for the movieWriter.

Obj-c code
`[movieWriter setCompletionBlock:^{
        [filter removeTarget:movieWriter];
        [movieWriter finishRecording];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [timer invalidate];
            self.progressLabel.text = @"100%";
        });
        
        UISaveVideoAtPathToSavedPhotosAlbum(pathToMovie, nil, NULL, NULL);
    }];`

How do you implement/translate this code in swift? Is the library ready for that?

**Comment by haraldmartin** [comment]
Not really an issue with the GPUImage source but anyway, blocks in Swift should be something like this:

```
movieWriter.setCompletionBlock({
    // …

    dispatch_async(dispatch_get_main_queue(), {
        // …
    })
})
```

**Comment by blaineam** [comment]
@ignasivegas I'm also trying to do the same thing to no avail. GPUImage fails each time it gets called.

this is the function i wrote so far for changing video filters
` func changevideofilter(filternumber: Int){

```
    switch filternumber {
    case 0:

        let documentsPath: NSString = NSSearchPathForDirectoriesInDomains(.DocumentDirectory,.UserDomainMask,true)[0] as NSString

        let importPath: NSString = documentsPath.stringByAppendingFormat("/xlargevideo.mov")
        let importUrl: NSURL = NSURL.fileURLWithPath(importPath as String)



        let asset = AVURLAsset(URL: importUrl)

        let videoAssetTrack = asset.tracksWithMediaType(AVMediaTypeVideo)[0]

        let movieFile = GPUImageMovie.init(URL: importUrl)
        movieFile.runBenchmark = false;
        movieFile.playAtActualSpeed = false;
        let filter = GPUImageToonFilter()
        movieFile.addTarget(filter)

        let filterView: GPUImageView = GPUImageView(frame: self.previewview.frame)
        filter.addTarget(filterView)
        // Only rotate the video for display, leave orientation the same for recording
       // GPUImageView *filterView = (GPUImageView *)self.view;
        //[filter addTarget:filterView];
        let exportPath: NSString = documentsPath.stringByAppendingFormat("/xlargevideofiltered.mov")
        let exportUrl: NSURL = NSURL.fileURLWithPath(exportPath as String)


        let fileManager = NSFileManager.defaultManager()

        // Delete 'hello.swift' file

        do {
            try fileManager.removeItemAtPath(exportUrl.relativePath! as String)
        }
        catch let error as NSError {
            print("Ooops! Something went wrong: \(error)")
        }




        let movieWriter = GPUImageMovieWriter.init(movieURL: exportUrl, size: CGSizeMake(videoAssetTrack.naturalSize.width, videoAssetTrack.naturalSize.height))



        // In addition to displaying to the screen, write out a processed version of the movie to disk

        //let movieWriter = GPUImageMovieWriter.init(movieURL: exportUrl, size: CGSizeMake(720, 1152))

        filter.addTarget(movieWriter)

        movieWriter.shouldPassthroughAudio = true
        movieWriter.encodingLiveVideo = false

        movieFile.audioEncodingTarget = movieWriter

        movieFile.enableSynchronizedEncodingUsingMovieWriter(movieWriter)


        movieWriter.startRecording()

        movieFile.startProcessing()

        movieWriter.completionBlock = ({
                // …

                dispatch_async(dispatch_get_main_queue(), {
                    filter.removeTarget(movieWriter)
                    movieWriter.finishRecording()
                    self.processVideo(exportUrl)
                })
            })

        break
    default:

        break
    }




}`
```

if gets called a completion block for a high resolution video file crop render from this snippet:
` let exporter = AVAssetExportSession(asset: asset, presetName: AVAssetExportPresetHighestQuality)
        exporter!.videoComposition = videoComposition
        exporter!.outputFileType = AVFileTypeQuickTimeMovie
        exporter!.outputURL = exportUrl

```
    //exporter!.shouldOptimizeForNetworkUse = true


    exporter!.exportAsynchronouslyWithCompletionHandler({

        //display video after export is complete, for example...
        //let outputURL:NSURL = exporter!.outputURL!;
        //let asset:AVURLAsset = AVURLAsset(URL: outputURL, options: nil)
        //let newPlayerItem:AVPlayerItem = AVPlayerItem(asset: asset)
        //self.changevideofilter(0)

        dispatch_async(dispatch_get_main_queue(), {
            self.changevideofilter(0)
        //self.processVideo((exporter?.outputURL)!)
        })
        //self.mPlayer = AVPlayer.playerWithPlayerItem(newPlayerItem) as AVPlayer

    })
```

`

it then is supposed to call a video compression function for uploading.

however it is creating a EXC_BAD_ACCESS error in the ProcessAsset function in GPUImageMovie.m.

I'd love to see a working swift example if you don't mind sharing it.

Thanks again

**Comment by apprisingapps** [comment]
was anybody able to solve this.  I am running into the same issue.  Thank

**Comment by khush004** [comment]
@techyowl Thank you for the code, Can you please tell me where is **processVideo** method? and I am getting crash please check bellowed Image Please Help.!

![screen shot 2018-11-24 at 11 50 45 am](https://user-images.githubusercontent.com/29862624/48965250-72f1e100-efdf-11e8-88ae-3b87b313ed84.png)

**Comment by blaineam** [comment]
Gave up over 3 years ago on this when I found a different github repo with functional code using native apple filters at the time for videos or photos. I don't remember where that other repo was or if it's even still online but it was tied to a YouTube tutorial video. This project would have been nice if it would have worked for me giving the more adjustable parameters and control over "filter" creation. My app has shifted into a hobby project after I acquired its assets due to a business closure. I just don't have time to trace down this bug for this old of an issue. Best suggestion would be find a different repo if your still having issues like I was unless the owner of this project has addressed these issues.
